modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientPrimaryReplicaTracker.java

```diff
@@ -0,0 +1,327 @@
+ *
+ * <p>Assumptions:
+ *     - Primary replicas are not changed often.
+ *     - We do "best effort" partition awareness - it is ok if we don't have the latest primary replicas at some point or don't have them
+ *       at all. What matters is that we have the correct assignment eventually.
+ *     - It is allowed to return incomplete assignment (null for some partitions) - better than nothing.
+ *     - We don't know which tables the client is going to use, so we track a common maxStartTime for all tables.
+ *
+ * <p>Tracking logic:
+ *     - Listen to election events from placement driver, update primary replicas. This is the main source of information.
+ *     - When we have not yet received events for all partitions of a certain table, and the client requests the assignment,
+ *       load it from the placement driver. Wait for a limited amount of time (in getPrimaryReplica) and return what we have.
+ *       Don't block the client for too long, it is better to miss the primary than to delay the request.
+ */
+public class ClientPrimaryReplicaTracker implements EventListener<EventParameters> {
+    private final ConcurrentHashMap<TablePartitionId, ReplicaHolder> primaryReplicas = new ConcurrentHashMap<>();
+
+    private final AtomicLong maxStartTime = new AtomicLong();
+
+    private final PlacementDriver placementDriver;
+
+    private final HybridClock clock;
+
+    private final CatalogService catalogService;
+
+    private final SchemaSyncService schemaSyncService;
+
+    /**
+     * Constructor.
+     *
+     * @param placementDriver Placement driver.
+     * @param catalogService Catalog.
+     * @param clock Hybrid clock.
+     * @param schemaSyncService Schema synchronization service.
+     */
+    public ClientPrimaryReplicaTracker(
+            PlacementDriver placementDriver,
+            CatalogService catalogService,
+            HybridClock clock,
+            SchemaSyncService schemaSyncService) {
+        this.placementDriver = placementDriver;
+        this.catalogService = catalogService;
+        this.clock = clock;
+        this.schemaSyncService = schemaSyncService;
+    }
+
+    /**
+     * Gets primary replicas by partition for the table.
+     *
+     * @param tableId Table ID.
+     * @param maxStartTime Timestamp.
+     * @return Primary replicas for the table, or null when not yet known.
+     */
+    public CompletableFuture<PrimaryReplicasResult> primaryReplicasAsync(int tableId, @Nullable Long maxStartTime) {
+        HybridTimestamp timestamp = clock.now();
+
+        if (maxStartTime == null) {
+            maxStartTime = this.maxStartTime.get();
+        } else {
+            // If the client provides an old maxStartTime, ignore it and use the current one.
+            maxStartTime = Math.max(maxStartTime, this.maxStartTime.get());
+        }
+
+        // Check happy path: if we already have all replicas, and maxStartTime > timestamp, return synchronously.
```

Comment:
`maxStartTime > timestamp` why do we need this? Seems that if methodParam.maxStartTime <= local.MaxStartTime is enough.

Reply:
> `methodParam.maxStartTime <= local.MaxStartTime`

Yes, that's what we check in `primaryReplicasNoWait`. The comment was a bit misleading, I've fixed it.